In [1]:
#!pip install tensorflow==2.0.0-rc1 --quiet

import warnings
warnings.filterwarnings("ignore")

from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential 
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.datasets import fashion_mnist
from sklearn.metrics import confusion_matrix, classification_report
import pandas as pd

# Data Cleaning

In [29]:
(X_train_raw, y_train), (X_test_raw, y_test) = fashion_mnist.load_data()

In [30]:
# scale
X_train = X_train_raw / 255.0
X_test = X_test_raw / 255.0

# reshape to use in Dense layers
X_train = X_train.reshape(X_train.shape[0], 28*28)
X_test = X_test.reshape(X_test.shape[0], 28*28)

In [16]:
X_test.shape

(10000, 784)

In [36]:
# convert labels to numeric
y_train_cat = to_categorical(y_train, 10)
y_test_cat = to_categorical(y_test, 10)

In [42]:
# want to keep y_test unaltered for use in evaluation
y_test

array([9, 2, 1, ..., 8, 1, 5], dtype=uint8)

In [41]:
y_test_cat

array([[0., 0., 0., ..., 0., 0., 1.],
       [0., 0., 1., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]], dtype=float32)

# Building the Models

## Model 1: 5 layers, Adam

In [43]:
model_1 = Sequential()
model_1.add(Dense(128, input_dim = X_train.shape[1], activation = 'relu'))
model_1.add(Dense(64, activation = 'relu'))
model_1.add(Dense(64, activation = 'relu'))
model_1.add(Dense(32, activation = 'relu'))
model_1.add(Dense(10, activation = 'softmax'))

In [44]:
model_1.compile(loss = 'categorical_crossentropy', optimizer = 'adam', metrics = ['accuracy'])

model_1.fit(X_train, y_train_cat, validation_data = (X_test, y_test_cat), 
         epochs = 100, batch_size = 200, verbose = 1)

Train on 60000 samples, validate on 10000 samples
Epoch 1/100
60000/60000 [==============================] - 3s 43us/sample - loss: 0.6216 - accuracy: 0.7818 - val_loss: 0.4759 - val_accuracy: 0.8347
Epoch 2/100
60000/60000 [==============================] - 2s 32us/sample - loss: 0.4130 - accuracy: 0.8548 - val_loss: 0.4207 - val_accuracy: 0.8457
Epoch 3/100
60000/60000 [==============================] - 2s 34us/sample - loss: 0.3669 - accuracy: 0.8683 - val_loss: 0.3856 - val_accuracy: 0.8596
Epoch 4/100
60000/60000 [==============================] - 2s 33us/sample - loss: 0.3343 - accuracy: 0.8790 - val_loss: 0.3843 - val_accuracy: 0.8621
Epoch 5/100
60000/60000 [==============================] - 2s 33us/sample - loss: 0.3151 - accuracy: 0.8846 - val_loss: 0.3662 - val_accuracy: 0.8675
Epoch 6/100
60000/60000 [==============================] - 2s 33us/sample - loss: 0.2975 - accuracy: 0.8915 - val_loss: 0.3453 - val_accuracy: 0.8738
Epoch 7/100
60000/60000 [=========================

Epoch 50/100
60000/60000 [==============================] - 2s 33us/sample - loss: 0.1065 - accuracy: 0.9602 - val_loss: 0.4976 - val_accuracy: 0.8900
Epoch 51/100
60000/60000 [==============================] - 2s 34us/sample - loss: 0.1014 - accuracy: 0.9618 - val_loss: 0.5030 - val_accuracy: 0.8924
Epoch 52/100
60000/60000 [==============================] - 3s 44us/sample - loss: 0.0942 - accuracy: 0.9645 - val_loss: 0.4732 - val_accuracy: 0.8887
Epoch 53/100
60000/60000 [==============================] - 3s 51us/sample - loss: 0.0947 - accuracy: 0.9640 - val_loss: 0.5416 - val_accuracy: 0.8870
Epoch 54/100
60000/60000 [==============================] - 3s 50us/sample - loss: 0.0911 - accuracy: 0.9649 - val_loss: 0.4978 - val_accuracy: 0.8871
Epoch 55/100
60000/60000 [==============================] - 2s 40us/sample - loss: 0.0896 - accuracy: 0.9658 - val_loss: 0.5298 - val_accuracy: 0.8868
Epoch 56/100
60000/60000 [==============================] - 2s 35us/sample - loss: 0.0884 - ac

Train Loss: 0.0.0416 <br>
Train Accuracy: 0.9846 <br>
Test Loss: 0.7687 <br>
Test Accuracy: 0.8884

## Model 2: 5 layers, RMSprop

In [45]:
model_2 = Sequential()
model_2.add(Dense(128, input_dim = X_train.shape[1], activation = 'relu'))
model_2.add(Dense(64, activation = 'relu'))
model_2.add(Dense(64, activation = 'relu'))
model_2.add(Dense(32, activation = 'relu'))
model_2.add(Dense(10, activation = 'softmax'))

In [46]:
model_2.compile(loss = 'categorical_crossentropy', optimizer = 'RMSprop', metrics = ['accuracy'])

model_2.fit(X_train, y_train_cat, validation_data = (X_test, y_test_cat), 
         epochs = 100, batch_size = 200, verbose = 1)

Train on 60000 samples, validate on 10000 samples
Epoch 1/100
60000/60000 [==============================] - 3s 51us/sample - loss: 0.6557 - accuracy: 0.7659 - val_loss: 0.5027 - val_accuracy: 0.8126
Epoch 2/100
60000/60000 [==============================] - 2s 36us/sample - loss: 0.4351 - accuracy: 0.8415 - val_loss: 0.4194 - val_accuracy: 0.8474
Epoch 3/100
60000/60000 [==============================] - 2s 35us/sample - loss: 0.3791 - accuracy: 0.8575 - val_loss: 0.4836 - val_accuracy: 0.8143
Epoch 4/100
60000/60000 [==============================] - 2s 34us/sample - loss: 0.3507 - accuracy: 0.8703 - val_loss: 0.3871 - val_accuracy: 0.8584
Epoch 5/100
60000/60000 [==============================] - 2s 38us/sample - loss: 0.3272 - accuracy: 0.8781 - val_loss: 0.3767 - val_accuracy: 0.8654
Epoch 6/100
60000/60000 [==============================] - 2s 40us/sample - loss: 0.3113 - accuracy: 0.8845 - val_loss: 0.3765 - val_accuracy: 0.8657
Epoch 7/100
60000/60000 [=========================

Epoch 50/100
60000/60000 [==============================] - 2s 34us/sample - loss: 0.1145 - accuracy: 0.9565 - val_loss: 0.5612 - val_accuracy: 0.8895
Epoch 51/100
60000/60000 [==============================] - 2s 34us/sample - loss: 0.1117 - accuracy: 0.9571 - val_loss: 0.6428 - val_accuracy: 0.8891
Epoch 52/100
60000/60000 [==============================] - 2s 34us/sample - loss: 0.1111 - accuracy: 0.9574 - val_loss: 0.5511 - val_accuracy: 0.8927
Epoch 53/100
60000/60000 [==============================] - 2s 35us/sample - loss: 0.1095 - accuracy: 0.9582 - val_loss: 0.5966 - val_accuracy: 0.8893
Epoch 54/100
60000/60000 [==============================] - 2s 36us/sample - loss: 0.1072 - accuracy: 0.9593 - val_loss: 0.6248 - val_accuracy: 0.8903
Epoch 55/100
60000/60000 [==============================] - 2s 35us/sample - loss: 0.1072 - accuracy: 0.9608 - val_loss: 0.5762 - val_accuracy: 0.8777
Epoch 56/100
60000/60000 [==============================] - 2s 36us/sample - loss: 0.1052 - ac

Train Loss: 0.0698 <br>
Train Accuracy: 0.9748 <br>
Test Loss: 1.0122 <br>
Test Accuracy: 0.8924 <br>

Using RMSprop as the optimizer very slightly increases test accuracy, but the loss is greater.

## Model 3: 5 layers, SGD

In [47]:
model_3 = Sequential()
model_3.add(Dense(128, input_dim = X_train.shape[1], activation = 'relu'))
model_3.add(Dense(64, activation = 'relu'))
model_3.add(Dense(64, activation = 'relu'))
model_3.add(Dense(32, activation = 'relu'))
model_3.add(Dense(10, activation = 'softmax'))

In [48]:
model_3.compile(loss = 'categorical_crossentropy', optimizer = 'sgd', metrics = ['accuracy'])

model_3.fit(X_train, y_train_cat, validation_data = (X_test, y_test_cat), 
         epochs = 100, batch_size = 200, verbose = 1)

Train on 60000 samples, validate on 10000 samples
Epoch 1/100
60000/60000 [==============================] - 2s 35us/sample - loss: 1.5913 - accuracy: 0.4693 - val_loss: 0.9736 - val_accuracy: 0.6493
Epoch 2/100
60000/60000 [==============================] - 2s 27us/sample - loss: 0.8093 - accuracy: 0.7107 - val_loss: 0.7368 - val_accuracy: 0.7381
Epoch 3/100
60000/60000 [==============================] - 2s 27us/sample - loss: 0.6658 - accuracy: 0.7682 - val_loss: 0.6511 - val_accuracy: 0.7695
Epoch 4/100
60000/60000 [==============================] - 2s 27us/sample - loss: 0.5925 - accuracy: 0.7960 - val_loss: 0.6240 - val_accuracy: 0.7704
Epoch 5/100
60000/60000 [==============================] - 2s 27us/sample - loss: 0.5481 - accuracy: 0.8095 - val_loss: 0.5487 - val_accuracy: 0.8110
Epoch 6/100
60000/60000 [==============================] - 2s 27us/sample - loss: 0.5191 - accuracy: 0.8191 - val_loss: 0.5274 - val_accuracy: 0.8160
Epoch 7/100
60000/60000 [=========================

Epoch 50/100
60000/60000 [==============================] - 2s 26us/sample - loss: 0.3013 - accuracy: 0.8913 - val_loss: 0.3710 - val_accuracy: 0.8699
Epoch 51/100
60000/60000 [==============================] - 2s 26us/sample - loss: 0.2991 - accuracy: 0.8946 - val_loss: 0.3611 - val_accuracy: 0.8722
Epoch 52/100
60000/60000 [==============================] - 2s 26us/sample - loss: 0.2970 - accuracy: 0.8935 - val_loss: 0.3701 - val_accuracy: 0.8685
Epoch 53/100
60000/60000 [==============================] - 2s 25us/sample - loss: 0.2963 - accuracy: 0.8940 - val_loss: 0.3722 - val_accuracy: 0.8675
Epoch 54/100
60000/60000 [==============================] - 2s 26us/sample - loss: 0.2938 - accuracy: 0.8949 - val_loss: 0.3592 - val_accuracy: 0.8747
Epoch 55/100
60000/60000 [==============================] - 2s 26us/sample - loss: 0.2924 - accuracy: 0.8944 - val_loss: 0.3655 - val_accuracy: 0.8730
Epoch 56/100
60000/60000 [==============================] - 2s 26us/sample - loss: 0.2906 - ac

Train Loss: 0.2309 <br>
Train Accuracy: 0.9159 <br>
Test Loss: 0.3416 <br>
Test Accuracy: 0.8831 <br>


SGD had the smallest test loss and is the least overfit, so I will investigate the effect of deeper layers using this optimizer.

## Model 4: 7 layers, SGD

In [49]:
model_4 = Sequential()
model_4.add(Dense(128, input_dim = X_train.shape[1], activation = 'relu'))
model_4.add(Dense(128, activation = 'relu'))
model_4.add(Dense(64, activation = 'relu'))
model_4.add(Dense(64, activation = 'relu'))
model_4.add(Dense(32, activation = 'relu'))
model_4.add(Dense(32, activation = 'relu'))
model_4.add(Dense(10, activation = 'softmax'))

In [50]:
model_4.compile(loss = 'categorical_crossentropy', optimizer = 'sgd', metrics = ['accuracy'])

model_4.fit(X_train, y_train_cat, validation_data = (X_test, y_test_cat), 
         epochs = 100, batch_size = 200, verbose = 1)

Train on 60000 samples, validate on 10000 samples
Epoch 1/100
60000/60000 [==============================] - 2s 38us/sample - loss: 1.7107 - accuracy: 0.4682 - val_loss: 1.0080 - val_accuracy: 0.6335
Epoch 2/100
60000/60000 [==============================] - 2s 29us/sample - loss: 0.8575 - accuracy: 0.6840 - val_loss: 0.7847 - val_accuracy: 0.7212
Epoch 3/100
60000/60000 [==============================] - 2s 30us/sample - loss: 0.7124 - accuracy: 0.7387 - val_loss: 0.6729 - val_accuracy: 0.7558
Epoch 4/100
60000/60000 [==============================] - 2s 31us/sample - loss: 0.6285 - accuracy: 0.7670 - val_loss: 0.6553 - val_accuracy: 0.7576
Epoch 5/100
60000/60000 [==============================] - 2s 32us/sample - loss: 0.5840 - accuracy: 0.7829 - val_loss: 0.6135 - val_accuracy: 0.7713
Epoch 6/100
60000/60000 [==============================] - 2s 36us/sample - loss: 0.5461 - accuracy: 0.7998 - val_loss: 0.5383 - val_accuracy: 0.8050
Epoch 7/100
60000/60000 [=========================

Epoch 50/100
60000/60000 [==============================] - 2s 29us/sample - loss: 0.2763 - accuracy: 0.8989 - val_loss: 0.3695 - val_accuracy: 0.8666
Epoch 51/100
60000/60000 [==============================] - 2s 30us/sample - loss: 0.2754 - accuracy: 0.8992 - val_loss: 0.3612 - val_accuracy: 0.8708
Epoch 52/100
60000/60000 [==============================] - 2s 30us/sample - loss: 0.2732 - accuracy: 0.9000 - val_loss: 0.3682 - val_accuracy: 0.8687
Epoch 53/100
60000/60000 [==============================] - 2s 31us/sample - loss: 0.2706 - accuracy: 0.9018 - val_loss: 0.3519 - val_accuracy: 0.8759
Epoch 54/100
60000/60000 [==============================] - 2s 31us/sample - loss: 0.2687 - accuracy: 0.9007 - val_loss: 0.3511 - val_accuracy: 0.8770
Epoch 55/100
60000/60000 [==============================] - 2s 30us/sample - loss: 0.2670 - accuracy: 0.9018 - val_loss: 0.3615 - val_accuracy: 0.8685
Epoch 56/100
60000/60000 [==============================] - 2s 29us/sample - loss: 0.2645 - ac

Train Loss: 0.1946 <br>
Train Accuracy: 0.9295 <br>
Test Loss: 0.3542 <br>
Test Accuracy: 0.8794 <br>

Increasing the number of layers decreased loss very slightly and actually decreased accuracy. It may not be worth the extra computation for such little increase in performance. The models are slightly overfit, so adding dropout layers might be helpful.

## Model 5: 5 layers, SGD, with Dropout

In [51]:
model_5 = Sequential()
model_5.add(Dropout(rate = 0.2, input_shape = (X_train.shape[1],)))
model_5.add(Dense(128, activation = 'relu'))
model_5.add(Dense(64, activation = 'relu'))
model_5.add(Dense(64, activation = 'relu'))
model_5.add(Dense(32, activation = 'relu'))
model_5.add(Dense(10, activation = 'softmax'))

In [52]:
model_5.compile(loss = 'categorical_crossentropy', optimizer = 'RMSprop', metrics = ['accuracy'])

model_5.fit(X_train, y_train_cat, validation_data = (X_test, y_test_cat), 
        epochs = 100, batch_size = 200, verbose = 1)

Train on 60000 samples, validate on 10000 samples
Epoch 1/100
60000/60000 [==============================] - 3s 53us/sample - loss: 0.6911 - accuracy: 0.7532 - val_loss: 0.5152 - val_accuracy: 0.8168
Epoch 2/100
60000/60000 [==============================] - 2s 38us/sample - loss: 0.4719 - accuracy: 0.8281 - val_loss: 0.4304 - val_accuracy: 0.8397
Epoch 3/100
60000/60000 [==============================] - 2s 39us/sample - loss: 0.4220 - accuracy: 0.8445 - val_loss: 0.4340 - val_accuracy: 0.8424
Epoch 4/100
60000/60000 [==============================] - 2s 39us/sample - loss: 0.3952 - accuracy: 0.8536 - val_loss: 0.3933 - val_accuracy: 0.8600
Epoch 5/100
60000/60000 [==============================] - 2s 38us/sample - loss: 0.3716 - accuracy: 0.8635 - val_loss: 0.3648 - val_accuracy: 0.8686
Epoch 6/100
60000/60000 [==============================] - 2s 39us/sample - loss: 0.3555 - accuracy: 0.8677 - val_loss: 0.4111 - val_accuracy: 0.8476
Epoch 7/100
60000/60000 [=========================

Epoch 50/100
60000/60000 [==============================] - 2s 38us/sample - loss: 0.2269 - accuracy: 0.9146 - val_loss: 0.3320 - val_accuracy: 0.8893
Epoch 51/100
60000/60000 [==============================] - 2s 38us/sample - loss: 0.2243 - accuracy: 0.9150 - val_loss: 0.3292 - val_accuracy: 0.8921
Epoch 52/100
60000/60000 [==============================] - 2s 38us/sample - loss: 0.2245 - accuracy: 0.9152 - val_loss: 0.3412 - val_accuracy: 0.8894
Epoch 53/100
60000/60000 [==============================] - 2s 39us/sample - loss: 0.2239 - accuracy: 0.9151 - val_loss: 0.3389 - val_accuracy: 0.8901
Epoch 54/100
60000/60000 [==============================] - 2s 38us/sample - loss: 0.2237 - accuracy: 0.9154 - val_loss: 0.3401 - val_accuracy: 0.8891
Epoch 55/100
60000/60000 [==============================] - 2s 38us/sample - loss: 0.2210 - accuracy: 0.9170 - val_loss: 0.3486 - val_accuracy: 0.8932
Epoch 56/100
60000/60000 [==============================] - 2s 39us/sample - loss: 0.2224 - ac

Train Loss: 0.2076 <br>
Train Accuracy: 0.9219 <br>
Test Loss: 0.4050<br>
Test Accuracy: 0.8921 <br>

Using a dropout layer decreased overfitting, and increased accuracy. Depending on the use case for this model, this could be acceptable. 

In [53]:
y_pred = model_5.predict_classes(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.89      0.76      0.82      1000
           1       0.99      0.98      0.98      1000
           2       0.80      0.80      0.80      1000
           3       0.86      0.94      0.90      1000
           4       0.78      0.86      0.82      1000
           5       0.99      0.97      0.98      1000
           6       0.74      0.70      0.72      1000
           7       0.95      0.97      0.96      1000
           8       0.96      0.97      0.97      1000
           9       0.96      0.97      0.96      1000

    accuracy                           0.89     10000
   macro avg       0.89      0.89      0.89     10000
weighted avg       0.89      0.89      0.89     10000



In [56]:
confusion_df = pd.DataFrame(confusion_matrix(y_test, y_pred),
                            index = ['T-shirt', 'Trouser', 'Pullover', 'Dress', 'Coat', 'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle Boot'],
                            columns = [' Pred T-shirt', 'Pred Trouser', 'Pred Pullover', 'Pred Dress', 'Pred Coat', 'Pred Sandal', 'Pred Shirt', 'Pred Sneaker', 'Pred Bag', 'Pred Ankle Boot'])

In [57]:
confusion_df

Pred T-shirt  Pred Trouser  Pred Pullover  Pred Dress  Pred Coat  \
T-shirt               761             3             21          44          4   
Trouser                 0           979              1          13          3   
Pullover                9             1            802          14        129   
Dress                   7             5             10         939         20   
Coat                    0             0             66          42        858   
Sandal                  0             0              0           1          0   
Shirt                  74             2             95          34         83   
Sneaker                 0             0              0           0          0   
Bag                     0             0              2           5          7   
Ankle Boot              0             0              0           0          0   

            Pred Sandal  Pred Shirt  Pred Sneaker  Pred Bag  Pred Ankle Boot  
T-shirt               0         151             0        16                0  
Trouser               0           2             0         2                0  
Pullover              0          43             0         2                0  
Dress                 0          15             0         4                0  
Coat                  0          33             0         1                0  
Sandal              967           0            14         0               18  
Shirt                 0         701             0        11                0  
Sneaker               5           0           971         0               24  
Bag                   3           5             4       974                0  
Ankle Boot            1           1            29         0              969

# Conclusion

The best model was 5 layers, used one dropout layer with a rate of 0.2, and a SGD optimizer. The model had the most difficulty classifiying class 4 and 6 (coats and shirts, respectively). Performance might be improved by adjusting the learning rate and momentum of the SGD optimizer.